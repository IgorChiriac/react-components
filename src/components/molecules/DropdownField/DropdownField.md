`<DropdownField />` has exactly the same behaviour as `<TextField />`, the only difference being that it renders a `<Dropdown />` instead of `<InputText />`.

```jsx
import { DropdownField } from '@zopauk/react-components';

<DropdownField
  label="Your cool dropdown ❤"
  errorMessage="You need to choose something!"
  size="short"
  name="foo"
  helpText="Some contextual help"
>
  <option value="first">First value</option>
  <option value="second">Second value</option>
  <option value="third">Third value</option>
</DropdownField>;
```
