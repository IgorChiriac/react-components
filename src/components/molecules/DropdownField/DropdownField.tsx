import React, { InputHTMLAttributes } from 'react';
import styled from 'styled-components';
import ErrorMessage from '../../atoms/ErrorMessage/ErrorMessage';
import Text from '../../atoms/Text/Text';
import InputLabel from '../../atoms/InputLabel/InputLabel';
import Dropdown from '../../atoms/Dropdown/Dropdown';
import SizedContainer from '../../layout/SizedContainer/SizedContainer';
import { IField } from '../../types';

interface IDropdownFieldProps
  extends Omit<IField, 'isValid' | 'inputProps'>,
    Omit<InputHTMLAttributes<HTMLSelectElement>, 'size'> {
  /**
   * Size attribute for the rendered HTML `<select>` element
   */
  htmlSelectSize?: InputHTMLAttributes<HTMLSelectElement>['size'];
}

const FieldError = styled(ErrorMessage)`
  margin-top: 5px;
`;

const Label = styled(InputLabel)`
  margin-bottom: 5px;
`;

function DropdownField(props: IDropdownFieldProps) {
  const { label, errorMessage, size, helpText, htmlSelectSize, name, ...rest } = props;

  if (!name) {
    throw Error("You didn't supply a name for the dropdown. Check the docs.");
  }

  const input = (
    <Dropdown id={`text-id-${name}`} aria-label={label ? undefined : name} {...rest} size={htmlSelectSize} />
  );

  return (
    <>
      {label && <Label htmlFor={`text-id-${name}`}>{label}</Label>}
      {helpText && <Text>{helpText}</Text>}
      <SizedContainer size={size}>{input}</SizedContainer>
      {errorMessage && <FieldError data-automation={`ZA.error-${name}`}>{errorMessage}</FieldError>}
    </>
  );
}

export default DropdownField;
