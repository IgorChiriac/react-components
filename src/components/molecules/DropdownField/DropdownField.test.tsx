import React from 'react';
import { render } from '@testing-library/react';
import { axe } from 'jest-axe';
import DropdownField from './DropdownField';

describe('<DropdownField />', () => {
  it('renders the component with props with no a11y violations', async () => {
    const { container } = render(
      <DropdownField
        label="Your cool dropdown ❤"
        errorMessage="You need to choose something!"
        size="short"
        name="dropdown-field-default"
      >
        <option value="first" key="first">
          First value
        </option>
        <option value="second" key="second">
          Second value
        </option>
        <option value="third" key="third">
          Third value
        </option>
      </DropdownField>,
    );

    const results = await axe(container.innerHTML);

    expect(container.firstChild).toMatchSnapshot();
    expect(results).toHaveNoViolations();
  });

  it('should throw if no name prop is provided', () => {
    expect(() =>
      render(
        <DropdownField
          label="Your cool dropdown ❤"
          errorMessage="You need to choose something!"
          size="short"
          name="dropdown-field-default"
        >
          <option value="first" key="first">
            First value
          </option>
        </DropdownField>,
      ),
    ).toThrowError("You didn't supply a name for the dropdown. Check the docs.");
  });
});
