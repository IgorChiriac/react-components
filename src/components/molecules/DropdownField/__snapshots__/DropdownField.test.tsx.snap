// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownField /> renders the component with props with no a11y violations 1`] = `
.c2 {
  margin: 0;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  line-height: 1.6;
  font-family: "Open Sans",-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  font-weight: 600;
  font-size: 16px;
}

.c1 {
  display: block;
}

.c0 {
  margin-bottom: 5px;
}

<label
  class="c0 c1 c2"
  for="text-id-dropdown-field-default"
>
  Your cool dropdown ‚ù§
</label>
`;
